docs/architecture.md

```diff
@@ -0,0 +1,215 @@
+2. **Reconcile**: When changes detected, run reconciliation loop
+3. **Converge**: Create/update Kubernetes resources to match desired state
+4. **Status Update**: Reflect observed state in Multigres status subresource
+5. **Requeue**: Schedule next reconciliation if needed
+
+### Component Reconcilers
+
+Each Multigres component has its own reconciler:
+
+- **etcd Reconciler**: Manages StatefulSet for etcd cluster, headless and client Services
+- **multigateway Reconciler**: Manages Deployment, Service, and optional HPA for MultiGateway
+- **multiorch Reconciler**: Manages Deployment and optional HPA for MultiOrch
+- **multipooler Reconciler**: Manages StatefulSet with multi-container pods (pooler, pgctld, postgres), and optional HPA
+
+All component reconcilers run in parallel.
+
+### Resource Builders
+
+Pure functions that generate Kubernetes manifests:
+
+- **Deterministic**: Same inputs always produce same outputs
+- **No Side Effects**: Don't make API calls or modify global state
+- **Testable**: Easily unit tested in isolation
+- **Composable**: Small functions that build specific resource types
+
+### Validation Strategy
+
+Using **CRD validation markers only** for simplicity:
+
+- OpenAPI v3 schema constraints (numeric ranges, enums, patterns, etc.)
+- Default values specified in CRD
+- API server enforces validation without external calls
+- No admission webhooks required
+
+### Observability
+
+**OpenTelemetry Integration**:
+- **Traces**: Reconciliation flow, API calls, component creation with spans
+- **Metrics**: Reconciliation duration, error rates, component health, resource counts
+- **Logs**: Structured logs with trace context correlation
+
+**Health Checks**:
+- **Liveness Probe**: HTTP endpoint to detect if operator needs restart
+- **Readiness Probe**: HTTP endpoint to indicate operator can handle requests
+
+**Kubernetes Events**:
+- Emitted for significant state changes and errors
+- Surfaced via `kubectl describe` for user visibility
+
+### Admission Webhooks (Future Consideration)
+
+**Current Decision**: Start without admission webhooks to keep installation simple and reduce moving parts during initial development.
+
+**When to Add**: Consider webhooks when:
+- Need validation beyond OpenAPI v3 schema (cross-field validation, complex business rules)
+- Want dynamic defaults based on cluster state
+- Need mutation beyond simple defaults
+
+**Certificate Management Options**:
+- **Init container pattern** (preferred): Jobs generate certs, init container waits - no runtime dependencies, standard CNCF pattern
+- **cert-manager**: Automatic cert management - adds runtime dependency but simplifies renewal
+- **Manual certificates**: Full control - operational overhead for rotation
+
+Current preference is init container pattern (same as Istio, NGINX Ingress), but final decision will be made during implementation based on operational requirements.
+
+## Technology Stack
+
+### Language and Runtime
+- **Language**: Go 1.24+
```

Comment:
also duplicated, recommend deduplicating.

Reply:
This one was on purpose, as I did want to touch on the very basic tech used, such as language. I think README helps anyone getting started, and this doc would be more in depth context that could be useful for in-depth understanding, as well as AI tooling.